torch/csrc/autograd/profiler.cpp

```diff
@@ -418,9 +485,62 @@ void Event::record(bool record_cuda) {
   cpu_ns_ = getTime();
 }
 
-double Event::cuda_elapsed_us(const Event & e) {
+/* static */ Event Event::fromIValue(const at::IValue& eventIValue) {
+  TORCH_INTERNAL_ASSERT(
+      eventIValue.isList(),
+      "Expected IValue to contain type c10::impl::GenericList");
+  auto ivalues = eventIValue.toList();
+  TORCH_INTERNAL_ASSERT(
+      ivalues.size() >= kEventIValuesSize,
+      "Expected at least ",
+      kEventIValuesSize,
+      " elements to reconstruct Event.");
+
+  Event evt(
+    static_cast<EventKind>(ivalues.get(0).toInt()), // EventKind
+    at::StringView(ivalues.get(1).toStringRef()), // name
+    ivalues.get(2).toInt(), // thread
+    static_cast<at::RecordFunctionHandle>(ivalues.get(3).toDouble()), // handle
+    {}, // TODO: record shapes
+    ivalues.get(4).toInt(), // node id
+    true, // is remote
+    ivalues.get(5).toInt(), // cpu_mem_usage
+    ivalues.get(6).toInt(), // cpu_ns
+    ivalues.get(7).toBool(), // was cuda recorded
+    ivalues.get(8).toInt(), // cuda memory usage
+    ivalues.get(9).toInt(), // device
+    ivalues.get(10).toDouble() // cuda_us
+  );
+  return evt;
+}
+
+at::IValue Event::toIValue() const {
+  c10::impl::GenericList eventIValueList(at::AnyType::get());
+  eventIValueList.reserve(kEventIValuesSize);
+  eventIValueList.emplace_back(static_cast<int64_t>(kind_));
+  eventIValueList.emplace_back(std::string(name_.str()));
+  eventIValueList.emplace_back(thread_id_);
+  eventIValueList.emplace_back(static_cast<double>(handle_));
+  eventIValueList.emplace_back(node_id_);
+  eventIValueList.emplace_back(cpu_memory_usage_);
+  eventIValueList.emplace_back(cpu_ns_);
+  // CUDA event information
+  bool cuda_profiling_enabled = has_cuda();
+  eventIValueList.emplace_back(cuda_profiling_enabled);
+  eventIValueList.emplace_back(static_cast<int64_t>(cuda_memory_usage_));
+  eventIValueList.emplace_back(device_);
+  eventIValueList.emplace_back(cuda_us_);
+  return at::IValue(eventIValueList);
+}
+
+double Event::cuda_elapsed_us(const Event & e) const {
+  if (isRemote() && e.isRemote()) {
+    // validate that cuda_us_ has been set properly.
+    TORCH_INTERNAL_ASSERT(cuda_us_ >= 0 && e.cuda_us_ >=0);
+    return e.cuda_us_ - cuda_us_;
+  }
   TORCH_CHECK(e.has_cuda() && has_cuda(), "Events were not recorded for CUDA");
```

Comment:
I guess that should be also checked for remote events as well

Reply:
Good point, even the next line device check should be the same

Reply:
Updated so that we do the below checks regardless of remote or not